<a href="https://colab.research.google.com/github/jugernaut/MACTI-analisisnumerico/blob/desarrollo/SegundoParcial/SegundoParcialNum.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<font color="Teal" face="Comic Sans MS,arial">
  <h1 align="center"><i>Segundo Parcial</i></h1>
  </font>
  <font color="Black" face="Comic Sans MS,arial">
  <h5 align="center"><i>Profesor: M.en.C. Miguel Angel Pérez León</i></h5>
  <h5 align="center"><i>Ayudante: Iván Calderón Coss</i></h5>
  <h5 align="center"><i>Materia: Análisis Numérico</i></h5>
</font>

# Nombre : Dereck Ariel Miguel Medrano
# No. cuenta: 318149875

## Las siguientes lineas son de apoyo para que ustedes sólo procesen los ejercicios

**Es necesario ejecutar las siguientes celdas, de otra forma no podrás realizar los ejercicios.**

In [ ]:
# bibliotecas de evaluacion automatizada
import numpy as np   
from macti.evaluacion import Evalua
evaluacion = Evalua('../../MACTI-analisisnumerico/SegundoParcial/utils/data/Parcial2', local=True)

# (5.25 Puntos) Segundo Parcial Parte Teórica

El examen consiste en elegir las preguntas que deseas responder para aprobar (en el mejor de los casos el examen completo).

En cada pregunta debes ser tan explícito como la calificación que desees obtener. En caso que tu justificación o desarrollo esté incompleto, también lo será la calificación asignada.

"Yo soy Thor, hijo de Odín y mientras haya vida en mi pecho… No se me ocurre más nada que decir. ¿Estás listo para morir?"

–Thor, la era de Ultrón.

# Normas Vectoriales


 ## (0.75 Puntos)
 
 Demuestra que la $\left\Vert \cdot\right\Vert _{\infty}$ es norma vectorial

Se debe demostrar que para cualesquiera vectores $\vec{x},\vec{y} \in \mathbb{R}^n$ se satisfacen las
siguientes propiedades.
1. $\left\Vert \vec{x}\right\Vert_{\infty} \geq0$\
Dem.
$$\text{Por def. de norma del supremo tenemos que: } $$
$$\left\Vert \vec{x}\right\Vert_{\infty} := sup\{|x_1|,|x_2|,...,|x_n|\}$$
donde para cada $i \in \{1,2,3..,n\}$
$$|x_i|\geq 0$$
y como cada entrada va ser postivia por elo valor absoluto, se cumple la propiedad.\
$$\therefore\left\Vert \vec{x}\right\Vert_{\infty} \geq0 $$

2.- $\left\Vert \vec{x}\right\Vert_{\infty} =0\,\Longleftrightarrow\vec{x}=\vec{0}$\
Dem.
$$\text{Por def. de norma del supremo tenemos que: } $$
$$\left\Vert \vec{x}\right\Vert_{\infty} := sup\{|x_1|,|x_2|,...,|x_n|\}$$
Donde\
$$sup\{|x_1|,|x_2|,...,|x_n|\} = 0$$
$$ \iff \forall i \in \{1,2,3..,n\}\text{   }  |x_i|= 0$$
$$\iff \vec{x}=\vec{0}$$
$$\therefore \left\Vert \vec{x}\right\Vert_{\infty} =0\,\Longleftrightarrow\vec{x}=\vec{0}$$


3.- $\left\Vert \alpha\vec{x}\right\Vert_{\infty} =\left|\alpha\right|\left\Vert \vec{x}\right\Vert_{\infty}$\
Dem.
$$\left\Vert \alpha\vec{x}\right\Vert_{\infty} =$$
$$ =sup\{|\alpha x_1|,|\alpha x_2|,...,|\alpha x_n|\}$$
Donde para algun $i \in \{1,2,3...,n\}$\
$$=|\alpha x_i| =|\alpha|| x_i| = |\alpha| sup\{|x_1|,|x_2|,...,|x_n|\}$$
$$=\left|\alpha\right|\left\Vert \vec{x}\right\Vert_{\infty} $$
$$\therefore\left\Vert \alpha\vec{x}\right\Vert_{\infty} =\left|\alpha\right|\left\Vert \vec{x}\right\Vert_{\infty}$$
4.- $\left\Vert \vec{x}+\vec{y}\right\Vert_{\infty} \leq \left\Vert \vec{x}\right\Vert_{\infty} +\left\Vert \vec{y}\right\Vert_{\infty}$\
Dem.
$$\left\Vert \vec{x}+\vec{y}\right\Vert_{\infty} =sup\{|x_1+y_1|,|x_2+y_2|,...,|x_n+y_n|\} $$
Aplicando desigualdad del triangulo del valor absoluto.\
$$\leq sup\{|x_1|+|y_1|,|x_2|+|y_2|,...,|x_n|+|y_n|\} $$
$$\leq sup\{|x_1|,|x_2|,...,|x_n|\} + sup\{|y_1|,|y_2|,...,|y_n|\}$$
$$=\left\Vert \vec{x}\right\Vert_{\infty} +\left\Vert \vec{y}\right\Vert_{\infty}$$
$$\therefore \left\Vert \vec{x}+\vec{y}\right\Vert_{\infty} \leq \left\Vert \vec{x}\right\Vert_{\infty} +\left\Vert \vec{y}\right\Vert_{\infty}$$

## (0.75 Puntos)

Sean $x\in\mathbb{R}^{n}$ y $\left\Vert \cdot\right\Vert$ Demostrar que la $\left\Vert \vec{x}\right\Vert _{\infty}$ y la $\left\Vert \vec{x}\right\Vert _{1}$ son equivalentes, es decir $\left\Vert \vec{x}\right\Vert _{\infty} \leq \left\Vert \vec{x}\right\Vert _{1}\leq n\left\Vert \vec{x}\right\Vert _{\infty}$. Se sugiere dividir la demostración en 2 partes.

*   $\left\Vert \vec{x}\right\Vert _{\infty} \leq \left\Vert \vec{x}\right\Vert _{1}$
*   $\left\Vert \vec{x}\right\Vert _{1}\leq n\left\Vert \vec{x}\right\Vert _{\infty}$


*   $\left\Vert \vec{x}\right\Vert _{\infty} \leq \left\Vert \vec{x}\right\Vert _{1}$\
Dem.
Sea $$\left\Vert \vec{x}\right\Vert_{\infty} := sup\{|x_1|,|x_2|,...,|x_n|\} = |x_j|$$
Para algun $j \in \{1,2,3,..,n\}$.\
Asi por def. de norma-1
$$\left\Vert \vec{x}\right\Vert _{1} = \sum_{i=1}^{n}\left|x_{i}\right| =\left|x_{1}\right|+\left|x_{2}\right|+\cdots + |x_j|+\cdots+\left|x_{n}\right| \geq |x_j| = $$
$$= sup\{|x_1|,|x_2|,...,|x_n|\} =\left\Vert \vec{x}\right\Vert_{\infty} $$
$$\therefore \left\Vert \vec{x}\right\Vert _{\infty} \leq \left\Vert \vec{x}\right\Vert _{1} $$


*   $\left\Vert \vec{x}\right\Vert _{1}\leq n\left\Vert \vec{x}\right\Vert _{\infty}$\
Dem.
$$|x_j|\geq |x_i| \implies \sum_{i=1}^{n}|x_j| \geq  \sum_{i=1}^{n}|x_i| $$
$$\implies n|x_j| \geq \left\Vert \vec{x}\right\Vert _{1}  $$
como $\left\Vert \vec{x}\right\Vert_{\infty} = |x_j| $
$$\therefore n\left\Vert \vec{x}\right\Vert _{\infty} \geq\left\Vert \vec{x}\right\Vert _{1} $$
Es decir.\
$$\left\Vert \vec{x}\right\Vert _{1}\leq n\left\Vert \vec{x}\right\Vert _{\infty}$$




# Normas Inducidas


## (0.75 Puntos) 

Derivar la formula para calcular $\left\Vert A\right\Vert _{\infty}$ (Teo. 2) explícitamente (seguir la derivación para $\left\Vert A\right\Vert _{1}$ vista en clase).

Dem.
Sea $\vec{x}\in\mathbb{R}^{n}$ y $A\in M_{n\times n}$ sobre $\mathbb{R}$ con $A=\left[\vec{a}_{1}|\vec{a}_{2}\text{|}\cdots\text{|}\vec{a}_{n}\right]$, donde $\vec{a}_{j}$ es un vector columna en $\mathbb{R}^{n}$. 
Consideremos la bola unitaria en la 1-norma $:=\left\{ \vec{x}\in\mathbb{R}^{n}\mid\left\Vert \vec{x}\right\Vert _{1}=\sum_{j=1}^{n}\left|x_{j}\right|\leq1\right\}$ . Cualquier vector $A\vec{x}$ satisface:

$$\left\Vert A\vec{x}\right\Vert _{\infty}	=
\left\Vert \sum_{j=1}^{n}\vec{a}_{j}x_{j}\right\Vert _{\infty}\overset{\underbrace{prop\,4\,nor}}{\leq}\sum_{j=1}^{n}\left\Vert \vec{a}_{j}x_{j}\right\Vert _{\infty} \leq \sum_{j=1}^{n}\left\Vert \vec{a}_{j}x_{j}\right\Vert _{1} = \sum_{j=1}^{n}\left\Vert \vec{a}_{j}\right\Vert _{1}|x_{j}|\leq\underset{1\leq i\leq n}{max} \left\Vert \vec{a}_{j}\right\Vert _{1} \sum_{j=1}^{n}|x_{j}| \overset{\underbrace{bola\,unit.}}{\leq} \underset{1\leq i\leq n}{max} \left\Vert \vec{a}_{j}\right\Vert _{1}= \underset{1\leq i\leq n}{max}\sum_{j=1}^{n}|a_{ij}|= \underset{1\leq i\leq n}{max}\left\Vert \vec{a}_{i}\right\Vert _{1}
$$

$$\therefore	\left\Vert A\vec{x}\right\Vert _{\infty}\leq\underset{1\leq i\leq n}{max}\left\Vert \vec{a}_{i}\right\Vert _{1}$$


Si elegimos $\vec{x}=\vec{e}_{i}$ (es decir el vector canónico, ademas $\left\Vert \vec{e}_{i}\right\Vert _{\infty}=1)$ con un 1 en la entrada $i$. Donde $i$ maximiza, $\left\Vert \vec{a}_{i}\right\Vert _{\infty}$, obtenemos la minima cota superior (en otras palabras el $sup$) y por lo tanto la igualdad.

$$\therefore\left\Vert A\right\Vert _{\infty}=\underset{\vec{x}\neq\vec{0}}{sup}\frac{\left\Vert A\vec{x}\right\Vert _{\infty}}{\left\Vert \vec{x}\right\Vert _{\infty}}=\underset{1\leq i\leq n}{max}\left\Vert \vec{a}_{j}\right\Vert _{1}$$

## (0.75 Puntos) 

Sea $A\in M_{n\times n}$ sobre $\mathbb{R}$ , demostrar las siguientes desigualdades de normas matriciales (inducidas).

*     $\left\Vert A\right\Vert _{1}\leq\sqrt{n}\left\Vert A\right\Vert _{2}$
*     $\sqrt{n}\left\Vert A\right\Vert _{2}\leq n\left\Vert A\right\Vert _{\infty}$

*Definición:* Dos normas $\left\Vert \cdot\right\Vert _{p}$ y $\left\Vert \cdot\right\Vert _{q}$
son equivalentes en $\mathbb{R}^{n}$ si existen 2 constantes universales
positivas $\alpha,\beta$ tales que

$$\forall\vec{x}\in\mathbb{R}^{n}\qquad\alpha\left\Vert \vec{x}\right\Vert _{p}\leq\left\Vert \vec{x}\right\Vert _{q}\leq\beta\left\Vert \vec{x}\right\Vert _{p}$$

$$\begin{eqnarray*}
1)\left\Vert \vec{x}\right\Vert _{2}\; & \leq & \left\Vert \vec{x}\right\Vert _{1}\leq\sqrt{n}\left\Vert \vec{x}\right\Vert _{2}\\
2)\left\Vert \vec{x}\right\Vert _{\infty} & \leq & \left\Vert \vec{x}\right\Vert _{2}\leq\sqrt{n}\left\Vert \vec{x}\right\Vert _{\infty}\\
3)\left\Vert \vec{x}\right\Vert _{\infty} & \leq & \left\Vert \vec{x}\right\Vert _{1}\leq n\left\Vert \vec{x}\right\Vert _{\infty}
\end{eqnarray*}$$

*     $\left\Vert A\right\Vert _{1}\leq\sqrt{n}\left\Vert A\right\Vert _{2}$\
Dem.\
$$\left\Vert A\right\Vert _{1} =  \underset{\vec{x}\neq\vec{0}}{sup}\frac{\left\Vert A\vec{x}\right\Vert _{1}}{\left\Vert \vec{x}\right\Vert _{1}}$$
Por $1)$
$$\leq \left\Vert A\right\Vert _{1} =  \underset{\vec{x}\neq\vec{0}}{sup}\frac{\sqrt{n}\left\Vert A\vec{x}\right\Vert _{2}}{\left\Vert \vec{x}\right\Vert _{1}}$$
Por $1)$
$$\left\Vert \vec{x}\right\Vert _{2} \leq  \left\Vert \vec{x}\right\Vert _{1} \iff \frac{1}{\left\Vert \vec{x}\right\Vert _{2}} \geq \frac{1}{\left\Vert \vec{x}\right\Vert _{1} }$$
$$\implies \left\Vert A\right\Vert _{1} =  \underset{\vec{x}\neq\vec{0}}{sup}\frac{\left\Vert A\vec{x}\right\Vert _{1}}{\left\Vert \vec{x}\right\Vert _{1}} \leq   \underset{\vec{x}\neq\vec{0}}{sup}\frac{\sqrt{n}\left\Vert A\vec{x}\right\Vert _{2}}{\left\Vert \vec{x}\right\Vert _{2}}= \sqrt{n}\left\Vert A\right\Vert _{2}$$
$$\therefore \left\Vert A\right\Vert _{1}\leq\sqrt{n}\left\Vert A\right\Vert _{2} $$




*     $\sqrt{n}\left\Vert A\right\Vert _{2}\leq n\left\Vert A\right\Vert _{\infty}$\
Dem.\

$$\sqrt{n}\left\Vert A\right\Vert _{2} = \underset{\vec{x}\neq\vec{0}}{sup}\frac{\sqrt{n}\left\Vert A\vec{x}\right\Vert _{2}}{\left\Vert \vec{x}\right\Vert _{2}}$$
Por $2)$ tenderemos que:
$$\left\Vert \vec{x}\right\Vert _{2} \leq \sqrt{n}\left\Vert \vec{x}\right\Vert _{\infty} \iff \sqrt{n}\left\Vert \vec{x}\right\Vert _{2} \leq n\left\Vert \vec{x}\right\Vert _{\infty} $$ 
$$\implies \sqrt{n}\left\Vert A\right\Vert _{2} = \underset{\vec{x}\neq\vec{0}}{sup}\frac{\sqrt{n}\left\Vert A\vec{x}\right\Vert _{2}}{\left\Vert \vec{x}\right\Vert _{2}} \leq \underset{\vec{x}\neq\vec{0}}{sup}\frac{n\left\Vert A\vec{x}\right\Vert _{\infty}}{\left\Vert \vec{x}\right\Vert _{2}}$$
Por $2)$
$$\left\Vert \vec{x}\right\Vert _{\infty}  \leq  \left\Vert \vec{x}\right\Vert _{2} \iff \frac{1}{\left\Vert \vec{x}\right\Vert _{\infty} } \geq \frac{1}{\left\Vert \vec{x}\right\Vert _{2}} $$

$$\implies \underset{\vec{x}\neq\vec{0}}{sup}\frac{n\left\Vert A\vec{x}\right\Vert _{\infty}}{\left\Vert \vec{x}\right\Vert _{2}} \leq \underset{\vec{x}\neq\vec{0}}{sup}\frac{n\left\Vert A\vec{x}\right\Vert _{\infty}}{\left\Vert \vec{x}\right\Vert _{\infty}} = n\left\Vert A\right\Vert _{\infty}$$

$$\therefore \sqrt{n}\left\Vert A\right\Vert _{2}\leq n\left\Vert A\right\Vert _{\infty}$$

## (0.75 Puntos) 
Sea $A=\left(\begin{array}{ccc}
1 & 5 & 8\\
4 & -5 & 10\\
9 & 15 & -2
\end{array}\right)$ calcular las siguientes normas (mostrar desarrollo):

*   $\left\Vert A\right\Vert _{1}$
*   $\left\Vert A\right\Vert _{\infty}$
*   $\left\Vert A\right\Vert _{2}$
*   $\left\Vert A\right\Vert _{F}$

### soluciones


*   $\left\Vert A\right\Vert _{1} = \underset{1\leq j\leq n}{max}\left\Vert \vec{a}_{j}\right\Vert _{1}=\underset{1\leq j\leq 3}{max}\left\Vert \vec{a}_{j}\right\Vert _{1}=max\{(1+4+9=14),(5+5+15=25),(8+10+2=20)\}=25$
*   $\left\Vert A\right\Vert _{\infty}=\underset{1\leq i\leq n}{max}\left\Vert \vec{a}_{i}\right\Vert _{1}=\underset{1\leq i\leq 3}{max}\left\Vert \vec{a}_{i}\right\Vert _{1}=max\{(1+5+8=14),(4+5+10=19),(9+15+2=26)\}=26$

Para las siguientes normas, primero calcularemos $(A^TA)$
Entonces:\
$$A^TA = \left(\begin{array}{ccc}
1 & 4 & 9\\
5 & -5 & 15\\
8 & 10 & -2
\end{array}\right)  \left(\begin{array}{ccc}
1 & 5 & 8\\
4 & -5 & 10\\
9 & 15 & -2
\end{array}\right) = \left(\begin{array}{ccc}
1+16+81 & 5-20+135 & 8+40-18\\
5-20+135 & 25+25+225 & 40-50-30\\
8+40-18 & 40-50-30 & 64+100+4
\end{array}\right) = \left(\begin{array}{ccc}
98 & 120 & 30\\
120 & 275 & -40\\
30 & -40 & 168
\end{array}\right)$$


*   $\left\Vert A\right\Vert _{2} = \sqrt{det\left(\left(A^{T}A\right)-\lambda I\right)}$
Eentonces, tenemos que:
$$\left(\begin{array}{ccc}
98 & 120 & 30\\
120 & 275 & -40\\
30 & -40 & 168
\end{array}\right) - \left(\begin{array}{ccc}
\lambda & 0 & 0\\
0 & \lambda & 0\\
0 & 0 & \lambda
\end{array}\right) = \left(\begin{array}{ccc}
98-\lambda & 120 & 30\\
120 & 275-\lambda & -40\\
30 & -40 & 168-\lambda
\end{array}\right)  $$

Asi,
$$det\left(\left(A^{T}A\right)-\lambda I\right) = 
det{\left(\begin{array}{ccc}
98-\lambda & 120 & 30\\
120 & 275-\lambda & -40\\
30 & -40 & 168-\lambda
\end{array}\right)} = 
98-\lambda \left(\begin{array}{cc}
275-\lambda & -40 \\
-40 & 168-\lambda 
\end{array}\right)
-120 \left(\begin{array}{cc}
120 & 30 \\
-40 & 168-\lambda 
\end{array}\right)+30\left(\begin{array}{cc}
120 & 30 \\
275-\lambda  & -40
\end{array}\right)
$$
$$=4370800-88014\lambda+541\lambda^2-\lambda^3 - 2563200+14400\lambda -391500+900\lambda$$
$$=1416100-72714\lambda+541\lambda^2-\lambda^3$$

Al encontrar las raíces de esta ecuación cubica, se tiene que$$\begin{cases}
\lambda_{1}=338.6057951\\
\lambda_{2}=179.034780.02566\\
\lambda_{3}=23.359423
\end{cases}$$

Donde si tomamos a $\lambda_{1}$\
$$\implies \left\Vert A\right\Vert _{2} = \sqrt{338.6057951}=18.40124$$

*   $\left\Vert A\right\Vert _{F}= \sqrt{traza\left(A^{T}A\right)} = \sqrt{98+275+168} = 23.2594067$

#### Comprobacion

In [ ]:
import numpy as np

# Se define la matriz A
A = np.array([[1,5,8],[4,-5,10],[9,15,-2]])
print (A)

# Calculamos la 1-norma de A
print (np.linalg.norm(A, 1))

# Calculamos la norma del supremo de A
print (np.linalg.norm(A, np.inf))

# Calculamos la 2-norma de A
print (np.linalg.norm(A, 2))

# Calculamos la norma F
print (np.linalg.norm(A, "fro"))

[[ 1  5  8]
 [ 4 -5 10]
 [ 9 15 -2]]
25.0
26.0
18.401244391160553
23.259406699226016


# Solución de Sistemas

## (0.5 Puntos)

Emplea algún tipo de factorización $\left(A=LU,\;PA=LU\right)$ para encontrar la solución del siguiente sistema de ecuaciones. Muestra el desarrollo, no solo el resultado.

$$\left(\begin{array}{ccc}
1 & 2 & 4\\
3 & 8 & 14\\
2 & 6 & 13
\end{array}\right)\left(\begin{array}{c}
x_{0}\\
x_{1}\\
x_{2}
\end{array}\right)=\left(\begin{array}{c}
3\\
13\\
4
\end{array}\right)$$

Sea:
$$A= \left(\begin{array}{ccc}
1 & 2 & 4\\
3 & 8 & 14\\
2 & 6 & 13
\end{array}\right)$$
$$\vec{b}= \left(\begin{array}{c}
3\\
13\\
4
\end{array}\right)$$
Aplicando la factorizacion $LU$ 
$$A=\left(\begin{array}{ccc}
1 & 2 & 4\\
3 & 8 & 14\\
2 & 6 & 13
\end{array}\right)=U\;y\;L=\left(\begin{array}{ccc}
1 & 0 & 0\\
0 & 1 & 0\\
0 & 0 & 1
\end{array}\right)\Longrightarrow A=UL$$
Multipliquemos el renglón 1 de $A$ por **3** y lo restamos al segundo renglón.
$$U=\left(\begin{array}{ccc}
1 & 2 & 4\\
0 & 2 & 2 \\
2 & 6 & 13
\end{array}\right)\quad L=\left(\begin{array}{ccc}
1 & 0 & 0\\
{\color{blue}{3}} & 1 & 0\\
0 & 0 & 1
\end{array}\right)$$
Multipliquemos el renglón 1 por **2** y lo restamos al tercer renglón.
$$U=\left(\begin{array}{ccc}
1 & 2 & 4\\
0 & 2 & 2 \\
0 & 2 & 5
\end{array}\right)\quad L=\left(\begin{array}{ccc}
1 & 0 & 0\\
3& 1 & 0\\
{\color{blue}{2}}  & 0 & 1
\end{array}\right)$$
Multipliquemos el renglón 2 por **1** y lo restamos al tercer renglón.
$$U=\left(\begin{array}{ccc}
1 & 2 & 4\\
0 & 2 & 2 \\
0 & 0 & 3
\end{array}\right)\quad L=\left(\begin{array}{ccc}
1 & 0 & 0\\
3& 1 & 0\\
 2 & {\color{blue}{1}}  & 1
\end{array}\right)$$

In [ ]:
U = np.array([[1,2,4],[0,2,2],[0,0,3]])
L = np.array([[1,0,0],[3,1,0],[2,1,1]])

# Comprobamos que la factorizacion es correcta
print(np.matmul(L,U))

[[ 1  2  4]
 [ 3  8 14]
 [ 2  6 13]]


Resolviendo $L\vec{y} = \vec{b}$
$$\left(\begin{array}{ccc}
1 & 0 & 0\\
3& 1 & 0\\
 2 & 1& 1
\end{array}\right) \left(\begin{array}{c}
y_0\\
y_1\\
y_2 
\end{array}\right) =\left(\begin{array}{c}
3\\
13\\
4
\end{array}\right) $$
Aplicancdo Substitucion hacia adelante.
$$
 \left(\begin{array}{c}
y_0\\
y_1\\
y_2 
\end{array}\right) 
 =  \left(\begin{array}{c}
3\\
13-3(y_0) = 4\\
4-y_1-2y_0=-6 
\end{array}\right) $$

Asi entonces, ahora tomamos $U\vec{x} = \vec{y}$
$$
\left(\begin{array}{ccc}
1 & 2 & 4\\
0 & 2 & 2 \\
0 & 0 & 3
\end{array}\right)\left(\begin{array}{c}
x_0\\
x_1\\
x_2 
\end{array}\right) = \left(\begin{array}{c}
3\\
4\\
-6 
\end{array}\right) $$
Aplicando substitucion hacia atras.
$$\left(\begin{array}{c}
x_0\\
x_1\\
x_2 
\end{array}\right)= \left(\begin{array}{c}
x_0 = 3 -4x_2-2x_1 = 3\\
x_1 = \frac{ 4-2x_2}{2}=4 \\
x_2 = \frac{-6}{3} =-2
\end{array}\right)$$
Por lo tanto la solucion al sistema es:
$$\left(\begin{array}{c}
x_0\\
x_1\\
x_2 
\end{array}\right)= \left(\begin{array}{c}
 3\\
4 \\
-2
\end{array}\right)$$

### Comprobacion

In [ ]:
import numpy as np

# algoritmo para sustitucion hacia delante
# n es el tamano de la dimension del problema
# matriz L, vector b ya estan dados como parametros
# guardar los resultados en el vector y
# Ly=b
def sustDelante(L, b):
    n=len(L)
    y=np.empty_like(b)
    y[0] = b[0]
    for i in range(1,n):
        y[i] = b[i]
        for j in range(0,i):
            y[i] -= L[i][j]*y[j]
    return y

# algoritmo para sustitucion hacia atras
# n es el tamano de la dimension del problema
# matriz U, vector y ya estan dados como parametros
# guardar los resultados en el vector x
# Ux=y
def sustAtras(U, y):
    n=len(U)
    x=np.empty_like(y)
    x[n-1] = y[n-1]/U[n-1][n-1]
    for i in range(n-2,-1,-1):
        x[i] = y[i]
        for j in range(i+1,n):
            x[i] -= U[i][j]*x[j]
        x[i] /= U[i][i]
    return x


In [ ]:
# Se definene los vectores
b = np.array([3,13,4])
# Usamos el algoritmo para encontrar la solucion
y = sustDelante(L, b)

# Se imprime el resultado
print(y)

[ 3  4 -6]


In [ ]:
# Usamos el algoritmo para encontrar la solucion
x = sustAtras(U, y)

# Se imprime el resultado
print(x)

[ 3  4 -2]


In [ ]:
A= np.array([[1,2,4],[3,8,14],[2,6,13]])
b = np.array([3,13,4])
#Usamos numpy para validar la solucion
print(np.linalg.solve(A,b))

[ 3.  4. -2.]


## (1 Punto) 

Calcula el número total de operaciones (en términos de $n$) para el algoritmo de sustitución hacia adelante. Muestra el desarrollo y concluye a que orden de complejidad pertenece este algoritmo, ¡Justifica tu respuesta!.

Usando el algoritmo de substitución:

$$y_{i}=\frac{b_{i}-\displaystyle\sum_{j=0}^{i-1}l_{ij}y_{j}}{l_{ii}}\quad i=0,\ldots,(n-1)$$
tenemos que:

| $$n$$        | $$i$$            |  $$operacion$$       | $$sumas$$|    $$Multiplicacoines$$|   $$Total$$| 
| :-:           |    :-:        |  :-:           |    :-:        |     :-:        |  :-:        | 
| $1$   | $0$     | $y_0 = \frac{b_0}{l_{00}}$ |$0$|$1$|$2$|
| $2$   | $1$     | $y_1 = \frac{b_1-l_{10}y_0}{l_{11}} $ |$1$|$2$|$3$| 
| $3$   | $2$     | $y_2 = \frac{b_2-l_{20}y_0-l_{21}y_1}{l_{22}}$ |$2$|$3$|$6$| 
| $4$   | $3$     | $y_3 =\frac{b_3-l_{30}y_0-l_{31}y_1-l_{32}y_2}{l_{33}}$ |$3$|$4$|$8$| 
| $5$   | $4$     | $y_4 = \frac{b_4-l_{40}y_0-l_{41}y_1-l_{42}y_2-l_{43}y_3}{l_{44}}$ |$4$|$5$|$10$| 
| $\cdots$  | $\cdots$    | $\cdots$ |$\cdots$ | $\cdots$ |$\cdots$ |
| $\infty$  | $n-1$      | $y_{i}=\frac{b_{i}-\displaystyle\sum_{j=0}^{i-1}l_{ij}y_{j}}{l_{ii}}$ |$n-1$ |$n$|$\frac{n(n-1)}{2}$|

 Recordadno que $"="$ la asigancion de valor es otra operacion, por eso e le agrega 1 al total

Con ello es orden de complejidad es de $\frac{n(n-1)}{2}$ el cual sabemos que pertenece $ O(n^{2})$
ya que:
$$\frac{n(n-1)}{2} = \frac{n^2-n}{2}=\frac{n^2}{2}-\frac{n}{2} \leq n^2$$ 
$$\therefore subadelante \in O(n^{2})$$

# (21 puntos) Segundo Parcial Parte Práctica

Ejecuta las celdas necesarias y valida tus resultados.

Observación: Las subsecciones se deben responder de manera consecutiva, ejemplo no se permite responder la subsección 1.2, sin antes haber respondido la subsección 1.1 y de igual manera con el resto de las subsecciones.

Newton multivariable: Considera el código del jupyter 'Sistemas no lineales', tal cómo se menciona en el mismo documento, el código que se muestra es obsoleto dado que no considera muchas características importantes, tales como: criterios de paro, evitar el cálculo de la inversa y aproximar las derivadas parciales. En cada subsección puedes hacer uso de las funciones de numpy para validar tus resultados.

La única subsección que otorga puntos parciales es la subsección 1.1, para cualquier otra subsección debe cumplirse con la totalidad con los puntos que se pide respectivamente y con la validación de resultados.

ESTRICTAMENTE PROHIBIDO HACER USO DE FUNCIONES QUE NO ESTEN PERMITIDAS.

In [ ]:
# Seccion 1
import numpy as np
import math

# funcion f1
def f1(x1, x2):
    return 2*x1-x2-math.e**(-x1)

# funcion f2
def f2(x1, x2):
    return -x1+2*x2-math.e**(-x2)

# denominador de la matriz jacobiana
def denominador(x1,x2):
    return 3+2*math.e**(-x2)+2*math.e**(-x1)+math.e**(-x1-x2)

# parcial (en este caso la parcial de f1 y f2 es la misma)
def par(x):
    return 2+math.e**(-x)

'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron(aprox):
    # contador de iteraciones
    n = 0 
    while n < 100: # maximo numero de iteraciones 100
        '''es necesario calcular la matriz jacobiana
        y para ello se necesita una matriz vacia de 2x2'''
        jacobinv = np.zeros([2,2])
        
        # valores de la matriz jacobiana en todas sus entradas
        jacobinv[0][0] = par(aprox[1])/denominador(aprox[0], aprox[1])
        jacobinv[0][1] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][0] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][1] = par(aprox[0])/denominador(aprox[0], aprox[1])
    
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox - np.matmul(jacobinv, fx)
        
        # se incrementa el contador
        n+=1
    
    # El valor devuelto es la aproximacion
    return aprox    
        
def main():
    #Aproximacion inicial (X=(0.0,0.0))
    ap = np.zeros([2])
    sol = JimyNewtron(ap)
    print('Aproximacion de la solucion')
    print(sol)
    print('Aproximacion evaluada en f1')
    print(f1(sol[0], sol[1]))
    print('Aproximacion evaluada en f2')
    print(f2(sol[0], sol[1]))

main()

Aproximacion de la solucion
[0.56714329 0.56714329]
Aproximacion evaluada en f1
-1.1102230246251565e-16
Aproximacion evaluada en f2
-1.1102230246251565e-16


## Sección 1.1 (2 puntos)

Modifica el código de forma tal que se contemplen los criterios de paro (Ecuaciones no lineales) vistos para métodos numéricos previos. **Unicamente se permite usar las funciones definidas en las subsección Tarea 4**.

*   Número máximo de iteraciones N.

*   Distancia de la evaluación de la solución menor o igual que la tolerancia $\left\Vert F\left(X^{k}\right)\right\Vert \leq Tol$.

*   Distancia entre la iteración actual $X^{k}$ y la iteración siguiente $X^{k+1}$ menor o igual que la tolerancia ($\left\Vert X^{k}-X^{k+1}\right\Vert \leq Tol$).

*   Utiliza las 3 normas definidas en la subsección Tarea 4 para los criterios de paro y fijate que sucede con la solución y con el número de iteraciones. Responde en en esta celda ¿Por qué los resultados y el número de iteraciones cambia al usar una norma diferente?.

In [ ]:
import numpy as np

def Norma1(vector):
    norma = 0
    for i in range(len(vector)):
        norma += abs(vector[i])
    return norma

def Norma2(vector):
    suma = 0
    for i in range(len(vector)):
        suma += abs(vector[i])**2
    norma = np.sqrt(suma)       
    return norma

def NormaInf(vector):
    vector_abs = []
    for i in range(len(vector)):
        vector_abs.append(abs(vector[i]))
    norma = max(vector_abs)
    return norma

In [ ]:
'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron1(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
        '''es necesario calcular la matriz jacobiana
        y para ello se necesita una matriz vacia de 2x2'''
        jacobinv = np.zeros([2,2])
        
        # valores de la matriz jacobiana en todas sus entradas
        jacobinv[0][0] = par(aprox[1])/denominador(aprox[0], aprox[1])
        jacobinv[0][1] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][0] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][1] = par(aprox[0])/denominador(aprox[0], aprox[1])
         
 
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
       
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        #Criterios de paro 
        if Norma1(fx)<=tol and Norma1(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox - np.matmul(jacobinv, fx)
        #print(aprox,"a")
        
        #print(fx,Norma1(fx),aprox, Norma1(aprox1-aprox))
        # se incrementa el contador
        n+=1
        

In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtron1(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))

evaluacion.verifica(np.array(round(sol[0],2),round(sol[1],2)),1)

Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07
¡Tu resultado es correcto!


### Usando las diferentes normas y su numero de iteraciones


### Norma-1

In [ ]:
'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtronN1(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
        '''es necesario calcular la matriz jacobiana
        y para ello se necesita una matriz vacia de 2x2'''
        jacobinv = np.zeros([2,2])
        
        # valores de la matriz jacobiana en todas sus entradas
        jacobinv[0][0] = par(aprox[1])/denominador(aprox[0], aprox[1])
        jacobinv[0][1] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][0] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][1] = par(aprox[0])/denominador(aprox[0], aprox[1])
         
 
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
       
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        #Criterios de paro 
        if Norma1(fx)<=tol and Norma1(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            print(f"numero de iteraciones: {n}")
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox - np.matmul(jacobinv, fx)
        #print(aprox,"a")
        
        print(fx,Norma1(fx),aprox, Norma1(aprox1-aprox))
        # se incrementa el contador
        n+=1
        

In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtronN1(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))

[-1. -1.] 2.0 [0.5 0.5] 1.0
[-0.10653066 -0.10653066] 0.21306131942526685 [0.566311 0.566311] 0.13262200639443633
[-0.00130451 -0.00130451] 0.0026090196120402975 [0.56714317 0.56714317] 0.0016643236752882373
numero de iteraciones3
Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07


### Norma-2

In [ ]:
'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtronN2(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
        '''es necesario calcular la matriz jacobiana
        y para ello se necesita una matriz vacia de 2x2'''
        jacobinv = np.zeros([2,2])
        
        # valores de la matriz jacobiana en todas sus entradas
        jacobinv[0][0] = par(aprox[1])/denominador(aprox[0], aprox[1])
        jacobinv[0][1] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][0] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][1] = par(aprox[0])/denominador(aprox[0], aprox[1])
         
 
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
       
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        #Criterios de paro 
        if Norma2(fx)<=tol and Norma2(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            print(f"numero de iteraciones: {n}")
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox - np.matmul(jacobinv, fx)
        #print(aprox,"a")
        
        #print(fx,Norma1(fx),aprox, Norma1(aprox1-aprox))
        # se incrementa el contador
        n+=1

In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtronN2(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))


numero de iteraciones: 3
Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07


### Norma-infinito

In [ ]:
'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron_inf(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
        '''es necesario calcular la matriz jacobiana
        y para ello se necesita una matriz vacia de 2x2'''
        jacobinv = np.zeros([2,2])
        
        # valores de la matriz jacobiana en todas sus entradas
        jacobinv[0][0] = par(aprox[1])/denominador(aprox[0], aprox[1])
        jacobinv[0][1] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][0] = 1/denominador(aprox[0], aprox[1])
        jacobinv[1][1] = par(aprox[0])/denominador(aprox[0], aprox[1])
         
 
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
       
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        #Criterios de paro 
        if NormaInf(fx)<=tol and NormaInf(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            print(f"numero de iteraciones: {n}")
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox - np.matmul(jacobinv, fx)
        #print(aprox,"a")
        
        #print(fx,Norma1(fx),aprox, Norma1(aprox1-aprox))
        # se incrementa el contador
        n+=1

In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtron_inf(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))


numero de iteraciones: 3
Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07


En conclusion, con cualquiera de las 3 normas, el numero de iteraciones sigue siendo el mismo, sin embargo, a la hora de calcular los valores de cada norma, se puede observar como por ejemplo la norma infinito se hace ma "pequeña" mas rapido, es decir, vuelve la distancia de las iteraciones menores o iguales a la tol, mas rapido.

## Sección 1.2 (+2 puntos)

En el código mostrado en la sección 1, se calcula la inversa del Jacobiano (lo cuál es un desperdicio de recursos). Modifica el código actual de la subsección 1.2, de tal manera que en lugar de calcular la inversa del Jacobiano, se resuelva el sistema lineal $\left[F'\left(\vec{X}\right)\right]H=-F\left(\vec{X}\right)$ (sistema lineal Jacobiano) para encontrar $H$ y modificar la solución actual, tal como se muestra en la sección del desarrollo en series de Taylor del jupyter [Sistemas no Lineales](../../MACTI-analisisnumerico/02_EcuacionesNoLineales/02_SistemasNoLineales.ipynb).

*    Para este inciso puedes usar el método $np.linalg.solve(\left[F'\left(\vec{X}\right)\right], -F\left(\vec{X}\right))$.



In [ ]:
# Seccion 1
import numpy as np
import math
#La funcion jacobiano, solo devuelve el jacobiano de la funcion ya definida
def jacobiano(x):
    return np.array([[par(x[0]), -1],[-1,par(x[1])]])

def Norma2(vector):
    suma = 0
    for i in range(len(vector)):
        suma += abs(vector[i])**2
    norma = np.sqrt(suma)       
    return norma


'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron2(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
      
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
        
        #Calculo del jacobiano
        Df = jacobiano(aprox)
        
        #Uso denumpy para resolver el sitema de ecuaciones
        H= np.linalg.solve(Df,-fx)
        
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        #Criterios de paro 
        if Norma2(fx)<=tol and Norma2(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            #print(f"numero de iteraciones: {n}")
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox + H
        #print(aprox,"a")
        
        #print(fx,Norma2(fx),aprox, Norma2(aprox1-aprox))
        # se incrementa el contador
        n+=1


In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtron2(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))

evaluacion.verifica(np.array(round(sol[0],2),round(sol[1],2)),2)

Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07
¡Tu resultado es correcto!


## Sección 1.3 (+3 puntos) 

Modifica el código de la subsección 1.2 y substituye el solver de numpy (`np.linalg.solve()`) por nuestros respectivos métodos. Utiliza los métodos vistos en clase:

*   `descomposicionLU`: factoriza una matriz en dos matrices L (lower) y U (upper).

*   `sustDelante`: algoritmo para la substitución hacia adelante.

*   `sustAtras`: algoritmo para la sustitución hacia atrás.

*   En suma, estos 3 métodos pueden substituir el método `np.linal.solve()` de *numpy*. De tal manera que ahora la forma de calcular $H$ se vea de la siguiente manera.
   
   `#H = np.linalg.solve(jacob, -fx) `         
   `H = solver(jacob, -fx)`
*   Donde `solver()` es la unión de los métodos antes descritos (`descomposicionLU`, `sustDelante`, `sustAtras`).

In [ ]:

import numpy as np
import math

# algoritmo para descomposicion A=LU
# la matriz A ya esta dada
def factorizacionLU(A):
    # dimension de la matriz
    n = len(A)
    # matriz L inicialmente es la identidad
    L = np.identity(n)
    # inicialmente la matriz A y la matriz U son iguales
    U = np.zeros((n,n))
    for i in range(0,n):
        for j in range(0,n):
            U[i][j] = A[i][j]
    # eliminacion gaussiana
    for i in range(0,n):
        for j in range(i+1,n):
            # guardar los factores de eliminacion gaussiana 
            # en la matriz L
            factor = U[j][i]/U[i][i]
            L[j][i] = factor
            # realizar eliminacion gaussiana en la matriz U
            # para quedar de forma triangular superior
            for k in range(i,n):
                U[j][k] = U[j][k] - factor*U[i][k]
    return L,U

# algoritmo para sustitucion hacia delante
# n es el tamano de la dimension del problema
# matriz L, vector b ya estan dados como parametros
# guardar los resultados en el vector y
# Ly=b
def sustDelante(L, b):
    n=len(L)
    y=np.empty_like(b)
    y[0] = b[0]
    for i in range(1,n):
        y[i] = b[i]
        for j in range(0,i):
            y[i] -= L[i][j]*y[j]
    return y

# algoritmo para sustitucion hacia atras
# n es el tamano de la dimension del problema
# matriz U, vector y ya estan dados como parametros
# guardar los resultados en el vector x
# Ux=y
def sustAtras(U, y):
    n=len(U)
    x=np.empty_like(y)
    x[n-1] = y[n-1]/U[n-1][n-1]
    for i in range(n-2,-1,-1):
        x[i] = y[i]
        for j in range(i+1,n):
            x[i] -= U[i][j]*x[j]
        x[i] /= U[i][i]
    return x

#Funcion solver, creada con el metodo LU y sustitucion
def solver(A,b):
    L,U=factorizacionLU(A)
    #Sustitucion adelante
    y = sustDelante(L, b)
    #sustitucion hacia atras
    resul = sustAtras(U, y)
    return resul
    

In [ ]:
'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron3(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
       
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
        
        #Calculo del jacobiano definido en la seccion anterior
        Df = jacobiano(aprox)
        #Uso de la funcion definida solver, con los metodos LU
        H = solver(Df,-fx)
        #H= np.linalg.solve(Df,-fx)
        
        aprox1 = aprox
        #print(aprox1,"a1")
        
        #print(Norma1(aprox1-aprox),"norma a1")
        
        
        #Criterios de paro 
        if Norma2(fx)<=tol and Norma2(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            #print(f"numero de iteraciones: {n}")
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox + H
        #print(aprox,"a")
        
        #print(fx,Norma2(fx),aprox, Norma2(aprox1-aprox))
        # se incrementa el contador
        n+=1


In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtron3(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))

evaluacion.verifica(np.array(round(sol[0],2),round(sol[1],2)),3)

Aproximacion de la solucion
[0.56714317 0.56714317]
Aproximacion evaluada en f1
-1.964804716703128e-07
Aproximacion evaluada en f2
-1.964804716703128e-07
¡Tu resultado es correcto!


## Sección 1.4 (+3 Puntos) 

Hasta este punto en el código de la subsección 1.3, las derivadas parciales empleadas para el cálculo del Jacobiano se definen en código duro (hard code).

*   Mediante el método:

`#parcial (en este caso la parcial de f1 y f2 es la misma) 
def par(x):     
	return 2+math.e**(-x)`

*   Modifica el código de la subsección 1.3 y encuentra alguna forma de aproximar las derivadas parciales, de tal manera que podamos aproximar las parciales de cualquier función $F\left(\vec{X}\right)$ con $\vec{X}\in\mathbb{R}^{n}$ y no tengamos que estar restringidos a emplear las parciales definidas de $f_{1}$ y $f_{2}$.

*   Sugerencia, el código debería verse similar a este.

`# F es una función vectorial
def parcial(F, X, var, delta):
devuelve la parcial de F en el punto X con respecto a la variable var usando delta`

*   Para la función parcial, basta con hacer uso de operaciones sobre listas de python y operaciones aritméticas elementales $\left(+,-,\times,\div\right)$. Para esta función solo se permite hacer uso de estas operadores aritméticos y funciones sobre listas (NO TE COMPLIQUES MAS DE LO NECESARIO).

In [ ]:
import numpy as np

def parcial(F, X, var, delta):
    """
    Calcula la derivada parcial de la función vectorial F en el punto X con respecto a la variable var
    utilizando un tamaño de paso delta.
    """
    # Copia el punto X para no modificarlo
    X_h = np.copy(X)
    
    # Aproxima la derivada parcial de F en X con respecto a la variable var 
    X_h[var] += delta
    F_h = F(X_h)
    X_h[var] -= delta
    F_x = F(X)
    df_dx = (F_h - F_x) / delta
    
    return df_dx

#Funcion vectorial
def F(X):
    x1, x2 = X
    return np.array([2*x1-x2-math.e**(-x1),-x1+2*x2-math.e**(-x2)])

# funcion f1
def f1(x1, x2):
    return 2*x1-x2-math.e**(-x1)

# funcion f2
def f2(x1, x2):
    return -x1+2*x2-math.e**(-x2)

'''Metodo de Newton para sistemas
   de ecuaciones no lineales
   aprox: es la primera aproximacion
   en forma de vector'''
def JimyNewtron4(aprox):
    # contador de iteraciones
    n = 0 
    #Numero maximo de iteraciones
    N=100
    #Tolerancia
    tol= 0.001
    while n < N: # maximo numero de iteraciones N
        
        J = np.zeros([2,2])
        # guarda la evalucion de f1 y f2 en forma de vector
        fx = np.array(aprox)
        fx[0] = f1(aprox[0], aprox[1])
        fx[1] = f2(aprox[0], aprox[1])
        
        #Calculo del jacobiano con las parciales
        for i in range(2):
            J[:, i] = parcial(F, aprox, i, tol)
            
        Df = J
        
        #Uso delsolver ya definido para encontrar la aproximacion
        H = solver(Df,-fx)
        
        aprox1 = aprox
        
        
        #Criterios de paro
        if Norma2(fx)<=tol and Norma2(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        aprox = aprox + H
        
        # se incrementa el contador
        n+=1


In [ ]:
#Aproximacion inicial (X=(0.0,0.0))
ap = np.zeros([2])
sol = JimyNewtron4(ap)
print('Aproximacion de la solucion')
print(sol)
print('Aproximacion evaluada en f1')
print(f1(sol[0], sol[1]))
print('Aproximacion evaluada en f2')
print(f2(sol[0], sol[1]))

evaluacion.verifica(np.array(round(sol[0],2),round(sol[1],2)),4)

[-1. -1.] 1.4142135623730951 [0.50012499 0.50012499] 0.7072835431467422
[-0.10632986 -0.10632986] 0.15037313696551965 [0.56632662 0.56632662] 0.09362324513829481
[-0.00128003 -0.00128003] 0.0018102315674915488 [0.56714332 0.56714332] 0.0011549831384622406
Aproximacion de la solucion
[0.56714332 0.56714332]
Aproximacion evaluada en f1
4.2460598703719654e-08
Aproximacion evaluada en f2
4.2460598703719654e-08
¡Tu resultado es correcto!


## Sección 1.5 (+10 Puntos)

Modifica el código de la subsección 1.4 de tal manera que la función `JimyNewtron` sea capaz de resolver cualquier sistema de n ecuaciones no lineales con n incognitas, no solamente uno de $2x2$ como lo hace actualmente. En otras palabras, actualmente `JimyNewtron` resuelve el siguiente sistema.

$$F\left(\vec{X}\right)=\begin{cases}
f_{1}\left(x_{1},x_{2}\right)\\
f_{2}\left(x_{1},x_{2}\right)
\end{cases}$$

Para obtener los 10 puntos de esta subsección debes modificar el código de la subsección 1.4, de tal manera que ahora sea capaz de resolver sistemas de n ecuaciones no lineales con n incognitas.

$$F\left(\vec{X}\right)=\begin{cases}
f_{1}\left(x_{1},x_{2},\ldots x_{n}\right)\\
f_{2}\left(x_{1},x_{2},\ldots x_{n}\right)\\
\cdots\\
f_{n}\left(x_{1},x_{2},\ldots x_{n}\right)
\end{cases}$$

Realiza tus pruebas y valida tus resultados.

## Validación

Antes de enviar tu examen valida tus resultados, **NO ENVIES CODIGO QUE NO HAYA SIDO VALIDADO**. En una celda de códio prueba tu algoritmo, es decir.

Con todas las secciones anteriores, ya podemos hacer uso de las funciones definidas, para poder calcular cualquier aproximacion con el metodo JimyNewtron para una funcion vectorial de n entradas.

In [ ]:
#Es necesario que se defina la funcion vecorial a la cual se le quiere calcular la aproximacion

#Funcion vectorial
def F1(X):
    x1,x2 = X
    return np.array([2*x1-x2-math.e**(-x1),-x1+2*x2-math.e**(-x2)])

#Calculo de la parcial de las entradas de la funcion
def parcial_vectorial(F, X, variable, delta):
    n = len(F(X))
    resultado = []
    try:
        i = int(variable)
    except ValueError:
        raise ValueError("La variable debe ser un índice entero.")
    X_delta = X.copy()
    X_delta[i] += delta
    for j in range(n):
        resultado.append((F(X_delta)[j] - F(X)[j]) / delta)
    return resultado

#Calculo del jacobiano de la funcion F, con la implementacion de la parcial
def jacobiangod(F, X, delta):
    # Calculamos el jacobiano de F en X
    n = len(X)
    J = np.zeros((n, n))
    for i in range(n):
        for j in range(n):
            J[i, j] = parcial_vectorial(F, X, j, delta)[i]
    return J

def solver(A,b):
    L,U=factorizacionLU(A)
    #Sustitucion adelante
    y = sustDelante(L, b)
    #sustitucion hacia atras
    resul = sustAtras(U, y)
    return resul

def JimyNewtron(aprox,tol,N):
  # contador de iteraciones
    n = 0 
    while n < N: # maximo numero de iteraciones N
 
        #Se evalua la funcion vectorial en la aprox
        FX=F1(aprox)
       
        #Se obtiene el jacobiano mediante la funcion jacobiangod
        #Dentro se calcula la parcial numerica respecto a las variables
        Df = jacobiangod(F1,aprox,tol)
     
        #Solver del sistema lineal de ecuaciones, para evitar el calculo de la inversa
        H = solver(Df,-FX)
        
        #Aproximacion k
        aprox1 = aprox
        
        #Condiciones de paro
        if Norma2(FX)<=tol and Norma2(aprox1-aprox) <= tol:
            # El valor devuelto es la aproximacion
            return aprox
        
        # FORMA ITERATIVA DEL METODO DE NEWTON PARA SISTEMAS NO LIENALES
        #  #Aproximacion k+1
        aprox = aprox + H
        # se incrementa el contador
        n+=1


In [ ]:
#Aproximacion inicial (X)     
ap = np.zeros([2])     
sol = JimyNewtron(ap, 0.001, 10000)     
print('Aproximacion de la solucion')     
print(sol)     
print('\nAproximacion evaluada en f1')     
print(F1(sol)) 

Aproximacion de la solucion
[0.56714332 0.56714332]

Aproximacion evaluada en f1
[4.24605987e-08 4.24605987e-08]


Sea cuál sea la subsección hasta la que hayas decidido responder (excepto la subseccion 1.5) la solución debe seguir siendo.

`Aproximacion de la solucion [0.56714332 0.56714332]`

`Aproximacion evaluada en F1 4.2460598703719654e-08`

`Aproximacion evaluada en F2 4.2460598703719654e-08`

# (¿? Puntos) Sección Especial

Esta sección es solo para amenizar al momento de calificar y probablemente subir unas cuantas décimas. Pero por ningún motivo esta sección hará que apruebes el examen si es que no muestras conocimientos.

“Tus antepasados lo llamaban magia, tú lo llamas ciencia. En mi mundo ambos son lo mismo.”

"La magia es sólo la ciencia que todavía no comprendemos."

 –Fráses de Thor, hijo de Odín.

<center>
<img src="https://github.com/jugernaut/ManejoDatos/blob/desarrollo/Imagenes/AlgoritmosBusqueda/traemethanos.gif?raw=1" width="600">   
</center> 

Thor Love and Thunder:
*   ¿Cómo se llama el martillo de Thor (escribe bien su nombre)?
*   ¿Cuántos hermanos tiene Thor?
*   ¿Con quien hace equipo Thor después de dejar de ser parte de los Vengadores
*   ¿Como se llama el antagonista de "Thor Love and Thunder" y por que desea matar dioses?
*   ¿Cuál es la última bendicion que recibe Thor y cómo se llama?

Thor Love and Thunder:
*   ¿Cómo se llama el martillo de Thor (escribe bien su nombre)?
mjolnir
*   ¿Cuántos hermanos tiene Thor?
*   ¿Con quien hace equipo Thor después de dejar de ser parte de los Vengadores
Guardianes de la galaxya
*   ¿Como se llama el antagonista de "Thor Love and Thunder" y por que desea matar dioses?
Gorr, y fue porque los dioses lo decepcionaron
*   ¿Cuál es la última bendicion que recibe Thor y cómo se llama?
